### PR TITLE
ENH: Allow performance warnings to be disabled

### LINKED ISSUE #56920: ENH: Allow users to disable PerformanceWarning
Part of the way toward #55385. Even if we decide not to implement #55385 (i.e. performance warnings are disabled by default), I think this is a good option for users to have.

I propose adding an option, `mode.performance_warnings` defaulting to True, that users could set to False to disable all PerformanceWarnings from being raised. The difference between this and #55385 is just changing the default value.

I opened #56921 to show the implementation.

Would like to get feedback on the general idea and happy to bikeshed the option name. Since this is impacts many places, cc @pandas-dev/pandas-core.

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #56920 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

~This still needs tests - I plan to add tests for disabling the warnings wherever we currently hit a PerformanceWarning in the test suite. But I'd like to get feedback on the linked issue first.~

Tests added.

### CODE REVIEW

**Comment by rhshadrach** [comment]
For the test fixture, I wanted to be able to pass it directly to `tm.assert_produces_warning` instead of having the fixture return True/False and then have lines like

    warn = PerformanceWarning if uses_performance_warning else False
    with tm.assert_produces_warning(warn):
        ...

File: pandas/core/arrays/arrow/_arrow_utils.py [hunk 1]
```diff
@@ -14,10 +16,11 @@ def fallback_performancewarning(version: str | None = None) -> None:
...
     """
-    msg = "Falling back on a non-pyarrow code path which may decrease performance."
-    if version is not None:
-        msg += f" Upgrade to pyarrow >={version} to possibly suppress this warning."
-    warnings.warn(msg, PerformanceWarning, stacklevel=find_stack_level())
+    if _get_option("performance_warnings"):
+        msg = "Falling back on a non-pyarrow code path which may decrease performance."
+        if version is not None:
+            msg += f" Upgrade to pyarrow >={version} to possibly suppress this warning."
+        warnings.warn(msg, PerformanceWarning, stacklevel=find_stack_level())
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/_arrow_utils.py:23]: Going forward how can we ensure that `PerformanceWarning`s are issued without checking the option?

Comment (rhshadrach) [reply pandas/core/arrays/arrow/_arrow_utils.py:23]: We currently raise if a warning is emitted in a test outside of `tm.assert_produces_warning`.

Comment (mroeschke) [reply pandas/core/arrays/arrow/_arrow_utils.py:23]: True, as long as we have a test for it :)

**Review by mroeschke — APPROVED** [review]
LGTM

**Comment by rhshadrach** [comment]
@mroeschke - should be good to merge.

**Comment by mroeschke** [comment]
Thanks @rhshadrach

**Comment by jorisvandenbossche** [comment]
Late feedback, but was thinking about this in the context of https://github.com/pandas-dev/pandas/issues/58581, but it would make sense to change the default to something like `None`? This default would mean that users see those performance warnings that we decided to turn on by default. More specifically, that would allow us to add new performance warnings that are _not_ shown by default (in contrast to existing ones), while still giving users a way to turn them on.

**Comment by rhshadrach** [comment]
That behavior sounds like an improvement to me, but I'd still prefer #55385 over it. Maybe discuss this alternative there? Would be good to have a list of performance warnings we'd always want on (I'd guess we'd start with whatever warnings we have now, but that might not be the same as want we want to end up with).

**Pull request merged**